### ISSUE TITLE
Add `epi_archive` class

### ISSUE THREAD

**Issue by brookslogan** [cmu-delphi/epiprocess#23]
Successor of issue #8.  Work in progress on PR #22.

Add a separate class to manage version history data for `epi_tibble`s, called `epi_archive` or `epi_tibble_archive`.  Make it work as outlined in [the issues vignette](https://github.com/cmu-delphi/epitools/blob/main/vignettes/issues.Rmd).  Complete the issues vignette to show this functionality.

**Comment by brookslogan** [comment]
See also [the table-diff-objects](https://github.com/cmu-delphi/epitools/tree/table-diff-objects) branch for work toward potentially useful/explanatory diff-of-data-table objects.  These may help compact the `update.DT` objects to only use a minimal number of rows, to validate user input, to update archives with additional data, to provide information for those modeling revisions, etc.

**Comment by brookslogan** [comment]
Additional tasks:
- [ ] the delphi-epidata issue data example may not match current epidata as-of query results when values are removed, although some PRs being integrated in covidcast will change them to agree.  This discrepancy should be noted.
- [ ] the delphi-epidata issue data example might be useful as its own set of functions for this package
- [ ] some functions to generate fake archive data for testing
- [ ] `testthat` tests

**Comment by brookslogan** [comment]
Old examples of epi_archive were removed at some point, but the second one was personally useful.  I'm just updating it a bit here to give the expected `TRUE` results on my current system (maybe different `all.equal` behavior?) and with the latest code.

```
library(dplyr)
## (delphi.epidata package is on GitHub in cmu-delphi/delphi-epidata-r)

update.df.2 =
  delphi.epidata::covidcast("jhu-csse", "confirmed_incidence_num",
                            "day", "state",
                            delphi.epidata::epirange(12340101,34560101), c("ak","al"),
                            issues = delphi.epidata::epirange(12340101,34560101)) %>%
  delphi.epidata::fetch_tbl()
snapshot.df.2a =
  delphi.epidata::covidcast("jhu-csse", "confirmed_incidence_num",
                            "day", "state",
                            delphi.epidata::epirange(12340101,34560101), c("ak","al"),
                            as_of = 20201014) %>%
  delphi.epidata::fetch_tbl()
snapshot.df.2b =
  delphi.epidata::covidcast("jhu-csse", "confirmed_incidence_num",
                            "day", "state",
                            delphi.epidata::epirange(12340101,34560101), c("ak","al"),
                            as_of = 20201028) %>%
  delphi.epidata::fetch_tbl()

epi.tibble.archive.2 = epiprocess::epi_archive$new(update.df.2)
all_equal(
  as_tibble(epi.tibble.archive.2$epi_df_as_of(as.Date("2020-10-14"))),
  as_tibble(snapshot.df.2a %>>% select(-issue)))
all_equal(
  as_tibble(epi.tibble.archive.2$epi_df_as_of(as.Date("2020-10-28"))),
  as_tibble(snapshot.df.2b %>>% select(-issue)))
```
Warnings are from #36 or something similar.

**Comment by brookslogan** [comment]
For preparation of "bleeding-edge" data frames with data as of t+k for every time t, here are a few approaches.  Looks like using the existing `update.DT` (+ an `update.lag` calculation in case there is no `lag` column in the input data) is comparable in speed to a more direct approach that would have required a different DT key.

```
update.df =
  delphi.epidata::covidcast("jhu-csse", "confirmed_incidence_num",
                            "day", "state",
                            delphi.epidata::epirange(12340101,34560101), "*",
                            issues = delphi.epidata::epirange(12340101,34560101)) %>%
  delphi.epidata::fetch_tbl()

lag.DT = as.data.table(update.df, key=c("geo_value", "time_value", "lag"))
k = 5

as.of.lag.k.a = unique(lag.DT[lag >= k], by=c("geo_value","time_value"), fromLast=FALSE)

update.DT = as.data.table(update.df, key=c("geo_value", "time_value", "issue"))
set(update.DT, , "update.lag", as.integer(update.DT[["issue"]] - update.DT[["time_value"]]))

as.of.lag.k.b = unique(update.DT[update.lag >= k], by=c("geo_value","time_value"), fromLast=FALSE)
set(as.of.lag.k.b, , "update.lag", NULL)

all_equal(as_tibble(as.of.lag.k.a), as_tibble(as.of.lag.k.b))

microbenchmark::microbenchmark(
                  unique(lag.DT[lag >= k], by=c("geo_value","time_value"), fromLast=FALSE),
                  unique(update.DT[update.lag >= k], by=c("geo_value","time_value"), fromLast=FALSE)[, update.lag := NULL][],
                  times=100L
                )
```

**Comment by ryantibs** [comment]
Finally closed!  Via #45
